### PR TITLE
feat: add support for inline struct creation

### LINKED ISSUE #2147: Support constructing new STRUCTs (i.e., writing nested data)
TL;DR: The goal is to support creating new STRUCTs in your queries.

# Background

KSQL already supports *reading* nested data (JSON, Avro) via the STRUCT type:

- Accessing (reading) existing nested data in your queries.  For example, you can access a nested field via the `myStruct->myField` syntax.
- "Limited write" support: you can copy nested data -- i.e., a STRUCT -- as-is (unmodified) from an input stream/table to an output/stream table.

However, KSQL does not yet support:

- Creating a new STRUCT from the input stream/table (i.e., in a CSAS/CTAS statement).

# Example

```sql
-- We start with a non-nested stream.
-- The goal is to turn the flat street and zip fields into a nested address STRUCT.
CREATE STREAM orders (item VARCHAR, quantity INTEGER, street VARCHAR, zip INTEGER) WITH (...);

-- The following is not yet supported!
-- Note: The syntax below is for illustrative purposes and does not preclude
-- the design discussion we need to have.
CREATE STREAM nested_orders AS
  SELECT
    item, quantity,
    STRUCT(street, zip) AS address1,
    STRUCT<renamedStreet VARCHAR, renamedZip INTEGER>(street, zip) AS address2,
    STRUCT<VARCHAR, LONG>(street, zip) AS address3
  FROM orders;
```

### PR DESCRIPTION
**Pull request by agavra**
fixes #2147 

### Description 
I was trying to debug an issue (#3874) but to my chagrin I was incapable of finding anyway to produce a struct with a decimal in it to a kafka topic. So I wrote this. What is this? This is the ability to create `STRUCT` types directly from KSQL - a long awaited feature! 

The syntax is up for debate (cc @hjafarpour @blueedgenick @rmoff) though I'd rather not go too far on that tangent as it's a long awaited feature. At the moment I chose ~`{(identifier expression (, identifier expression)*)?}`~ `STRUCT((expression AS identifier (, expression AS identifier)*)?`

Here's how it's used (note that the examples use the old syntax):

**INSERT VALUES:**
```sql
ksql> CREATE STREAM decimals (col0 STRUCT<val DECIMAL(2,1)>) WITH (kafka_topic='decimals', partitions=1, value_format='AVRO');

 Message
----------------
 Stream created
----------------
ksql> INSERT INTO decimals (col0) VALUES ({VAL '2.1'});
ksql> SELECT COL0->val FROM DECIMALS EMIT CHANGES;
+----------------------------------------------------------------------------------------------+
|COL0__VAL                                                                                     |
+----------------------------------------------------------------------------------------------+
|2.1                                                                                           |
```

**CSAS:**
```sql
ksql> CREATE STREAM decimals2 as SELECT {times2 col0->val * 2, times3 col0->val * 3} FROM DECIMALS;

 Message
-----------------------------------------------------------------------------------------
 Stream DECIMALS2 created and running. Created by query with query ID: CSAS_DECIMALS2_12
-----------------------------------------------------------------------------------------
```
```sql
ksql> SELECT KSQL_COL_0->TIMES2, KSQL_COL_0->TIMES3 FROM DECIMALS2 EMIT CHANGES;
+----------------------------------------------+----------------------------------------------+
|KSQL_COL_0__TIMES2                            |KSQL_COL_0__TIMES3                            |
+----------------------------------------------+----------------------------------------------+
|4.2                                           |6.3                                           |
```
```sql
ksql> DESCRIBE DECIMALS2;

Name                 : DECIMALS2
 Field      | Type
-----------------------------------------------------
 ROWTIME    | BIGINT           (system)
 ROWKEY     | VARCHAR(STRING)  (system)
 KSQL_COL_0 | STRUCT<TIMES3 DECIMAL, TIMES2 DECIMAL>
-----------------------------------------------------
For runtime statistics and query details run: DESCRIBE EXTENDED <Stream,Table>;
```

### Implementation

Ah... so you care about how it's implemented, eh? It's actually surprisingly similar to the `FunctionCall` and `SubscriptExpression` implementations combined:

- The `ExpressionTypeManager` can figure out the type of the struct by recursively visiting each field's expression
- The `CodeGenRunner` creates the schema once, and passes it as an argument to the `ExpressionEvaluator`
- The `SqlToJavaVisitor` uses the param that was passed in (see the `SqlToJavaVisitorTest` for an example) to construct the struct in the code that it generates
- The `SqlValueCoercer` helps with some edge cases in `INSERT VALUES` clauses (namely to insert decimals)

That's all folk!

### Testing done 

- Unit tests
- QTT Testing
- E2E Integration testing

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs-md/developer-guide/syntax-reference.md [hunk 1]
```diff
@@ -281,6 +281,12 @@ statement by using the syntax `STRUCT<FieldName FieldType, ...>`. For
 example, `STRUCT<ID BIGINT, NAME STRING, AGE INT>` defines a struct with
 three fields, with the supplied name and type.
 
+You can create a struct in a query by specifying the names of the columns
+and expressions that construct the values, separated by `,` and wrapped with
+curly braces. For example: `SELECT {name col0, ageInDogYears col1*7} AS dogs FROM animals`
+would create a schema `col0 STRUCT<VARCHAR, INTEGER>` (assuming `col0` was a string and
```
Comment (JimGalasyn) [inline docs-md/developer-guide/syntax-reference.md]:
```suggestion
creates a schema, `col0 STRUCT<VARCHAR, INTEGER>`, assuming `col0` is a string and
```

File: docs-md/developer-guide/syntax-reference.md [hunk 2]
```diff
@@ -281,6 +281,12 @@ statement by using the syntax `STRUCT<FieldName FieldType, ...>`. For
...
+`col1` was an integer).
```
Comment (JimGalasyn) [inline docs-md/developer-guide/syntax-reference.md]:
```suggestion
`col1` is an integer.
```

**Comment by blueedgenick** [comment]
Trying not to bikeshed this too hard - but you asked for feedback, so here it is 😉 

I'm not generally a fan of adding any extra syntax which depends on additional punctuation characters (the curly braces in this case) - it's always confusing trying to remember which brace-type to use in which situation and then you have to go look it up. Secondly, we have a (not uncommon) pattern of using functions to create non-primitive types (cf. `as_array()` and `as_map()`) - seems like we could follow that here too ? `as_struct()` looks like a candidate :) 

**Side note**: would be great if we fixed the parser and udf code to allow keywords as function names too, so we could rename these to just `array()`, `map()` and `struct()` but i guess that's for another day...

The difference with existing fns which create non-primitives is, of course, that we have to name the components of a struct. I think rather than `(identifier expression (, identifier expression)*)?` we could follow something like `(expression (as identifier)?, (expression (as identifier)?)*)?` instead - again for consistency with how we assign names to new attributes in projections. 
In conclusion, I'm leaning towards something like this in usage: `select as_struct(foo as bar, abc * 3) as my_new_struct_col from ...;`. Note that this also leaves open the option to not explicitly name the members of the new struct and have them auto-provided (KSQL_COL_0... etc) as necessary (as it would be for the result of `abc * 3` in the example).

File: ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4 [hunk 3]
```diff
@@ -257,6 +257,7 @@ primaryExpression
     | CASE whenClause+ (ELSE elseExpression=expression)? END                         #searchedCase
     | CAST '(' expression AS type ')'                                                #cast
     | ARRAY '[' (expression (',' expression)*)? ']'                                  #arrayConstructor
+    | '{' (identifier expression (',' identifier expression)*)? '}'                  #structConstructor
```
Comment (big-andy-coates) [inline ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4]: Like the array constructor on the line above, I would propose something more along the lines of:

```suggestion
    | STRUCT '(' expression (AS? identifier)? (',' expression (AS? identifier))* ')'                  #structConstructor
```

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java [hunk 4]
```diff
@@ -77,14 +83,20 @@ public void resolve(GenericRow row, Object[] parameters) {
     }
   }
 
+  public String getSchemaName(StructExpression structExpression) {
+    return structToCodeName.get(structExpression);
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: throw exception if not found, rather than returning `null`.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java [hunk 5]
```diff
@@ -77,14 +83,20 @@ public void resolve(GenericRow row, Object[] parameters) {
...
   static class Builder {
 
     private final ImmutableList.Builder<ArgumentSpec> argumentBuilder = ImmutableList.builder();
     private final Map<ColumnRef, String> columnRefToName = new HashMap<>();
     private final ImmutableListMultimap.Builder<FunctionName, String> functionNameBuilder =
         ImmutableListMultimap.builder();
+    private final Map<StructExpression, String> structToSchemaName = new HashMap<>();
 
     private int argumentCount = 0;
+    private int schemaCount = 0;
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]:
```suggestion
    private int structConstructorCount = 0;
```

File: ksql-execution/src/main/java/io/confluent/ksql/execution/expression/tree/StructExpression.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+import com.google.common.collect.ImmutableMap;
+import com.google.errorprone.annotations.Immutable;
+import io.confluent.ksql.parser.NodeLocation;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+@Immutable
+public class StructExpression extends Expression {
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/expression/tree/StructExpression.java]: I find this name a bit vague, would `StructConstructorExpression` not be a more descriptive name?

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java [hunk 7]
```diff
@@ -102,11 +114,18 @@ void addFunction(FunctionName functionName, Kudf function) {
       argumentBuilder.add(new FunctionArgumentSpec(codeName, function.getClass(), function));
     }
 
+    void addSchema(StructExpression struct, Schema schema) {
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: likewise, `addStructConstructor` is more descriptive than `addSchema` IMHO.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java [hunk 8]
```diff
@@ -102,11 +114,18 @@ void addFunction(FunctionName functionName, Kudf function) {
...
+      final String schemaName = CodeGenUtil.schemaName(schemaCount++);
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: do you need the counter? Can you not just use the structToSchemaName size?

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java [hunk 9]
```diff
@@ -102,11 +114,18 @@ void addFunction(FunctionName functionName, Kudf function) {
...
+      structToSchemaName.put(struct, schemaName);
+      argumentBuilder.add(new SchemaArgumentSpec(schemaName, schema));
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: Why is everything using the name `Schema`?  Isn't this specific to structs?

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java [hunk 10]
```diff
@@ -77,14 +83,20 @@ public void resolve(GenericRow row, Object[] parameters) {
...
+    return structToCodeName.get(structExpression);
+  }
+
   static class Builder {
 
     private final ImmutableList.Builder<ArgumentSpec> argumentBuilder = ImmutableList.builder();
     private final Map<ColumnRef, String> columnRefToName = new HashMap<>();
     private final ImmutableListMultimap.Builder<FunctionName, String> functionNameBuilder =
         ImmutableListMultimap.builder();
+    private final Map<StructExpression, String> structToSchemaName = new HashMap<>();
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: Other fields are using ImmutableXXX builders - can we follow the pattern?

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java [hunk 11]
```diff
@@ -208,4 +227,32 @@ public String toString() {
...
+  public static final class SchemaArgumentSpec extends BaseArgumentSpec {
+
+    private final Schema schema;
+
+    SchemaArgumentSpec(
+        String name,
+        Schema schema
+    ) {
+      super(name, Schema.class);
+      this.schema = schema;
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: nit: null check

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenUtil.java [hunk 12]
```diff
@@ -28,6 +29,10 @@ public static String paramName(int index) {
     return PARAM_NAME_PREFIX + index;
   }
 
+  public static String schemaName(int index) {
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenUtil.java]: why 'schema'?  Surely... 'struct'?

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java [hunk 13]
```diff
@@ -712,6 +722,25 @@ private String buildSupplierCode(String typeString, String code) {
...
+        struct.append(".put(")
+            .append('"')
+            .append(field.getKey())
+            .append('"')
+            .append(",")
+            .append(process(field.getValue(), context).getLeft())
+            .append(")");
+      }
+      return new Pair<>(
+          "((Struct)" + struct.toString() + ")",
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java]: think the cast is superfluous: 
```suggestion
          "(" + struct.toString() + ")",
```

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 14]
```diff
@@ -320,6 +323,19 @@ public Void visitSubscriptExpression(
       return null;
     }
 
+    @Override
+    public Void visitStructExpression(StructExpression exp, ExpressionTypeContext context) {
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: Be good to add a test with a nested struct constructor.

File: ksql-functional-tests/src/test/resources/query-validation-tests/create-struct.json [hunk 15]
```diff
@@ -0,0 +1,20 @@
+{
+  "comments": [
+    "Tests covering the use of the array returning UDFs."
+  ],
+  "tests": [
+    {
+      "name": "entries sorted",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/create-struct.json]: Can we add a nested struct creation test please?

Comment (big-andy-coates) [reply ksql-functional-tests/src/test/resources/query-validation-tests/create-struct.json]: And another that sees what happens if you try to create a struct with no fields...

File: ksql-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java [hunk 16]
```diff
@@ -77,6 +80,34 @@
...
+    final Struct struct = (Struct) value;
+    final Struct coerced = new Struct(
+        SchemaConverters.sqlToConnectConverter().toConnectSchema(targetType)
+    );
+
+    for (Field field : coerced.schema().fields()) {
+      Optional<io.confluent.ksql.schema.ksql.types.Field> sqlField = targetType.field(field.name());
+      if (!sqlField.isPresent()) {
+        // if there was a field in the struct that wasn't in the schema
+        // we should throw an exception
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java]: nit: we don't know the caller will throw an exception on `Optional.empty`.  I'd just say 'if there is a field in the struct that isn't in the schema we can not coerce'

File: ksql-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java [hunk 17]
```diff
@@ -77,6 +80,34 @@
...
+        return Optional.empty();
+      } else if (struct.schema().field(field.name()) == null) {
+        // if we cannot find the field in the struct, we can ignore it
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java]: Seems dangerous...

File: ksql-execution/src/main/java/io/confluent/ksql/execution/expression/tree/StructExpression.java (hunk 6 above)
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/expression/tree/StructExpression.java]: Changed to `CreateStructExpression`

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java (hunk 5 above)
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: The distinction between `struct` and `schema` is intentional in the code. The expression evaluator takes in as an argument the _schemas_ for the structs, not the struct expressions themselves. If you look at the code, it is used as `new Struct(schema0)...`. In this code, `schema0` is very much a schema.

It's just like how we have the method `addFunction` and not `addFunctionName`. The `functionName` is the key, but the `function` is what is actually being stored.

Resolving the other conversations that are about this to concentrate on this one.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java (hunk 8 above)
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: After changing this to use `ImmutableMap.Builder`, you can't get the size from the builder. I have kept this in favor of using ImmutableMap.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java (hunk 13 above)
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java]: I'm following the pattern in the rest of the code base here so I'll leave it as is

File: ksql-functional-tests/src/test/resources/query-validation-tests/create-struct.json (hunk 15 above)
Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/create-struct.json]: second test is in `insert-values` but I'll add one here too. Good call with nested structs.

File: ksql-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java (hunk 17 above)
Comment (agavra) [reply ksql-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java]: Oops! That's not what i meant to do... I meant to allow `STRUCT()` as an insert into statement for `STRUCT<foo VARCHAR>` but I don't think this condition is ever hit - there's a test for that somewhere. I'll do some more testing and get back to you.

**Comment by agavra** [comment]
Thanks for the reviews @big-andy-coates and @blueedgenick. Glad I asked for syntax feedback :) I've changed it to be `STRUCT((expression AS identifier (, expression AS identifier)*)?)`. I agree this feels more SQLy.

@blueedgenick - I didn't actually implement this with the UDF framework, so I could change the name to be exactly `STRUCT()` instead of `AS_STRUCT`. I might change array creation and map creation to be implemented inline. I don't think the UDF framework is necessarily the right place to be doing those type of type construction because schema inference is limited when you have arbitrary code generation and java generics. 

With regards to implicit naming (omitting the `AS`, that's backwards compatible to add later so I will do that in a future PR. I didn't want to think about edge cases (i.e. what happens if I create nested structs with the same names?) in this PR - though I don't think there are any problems.

@big-andy-coates - about decimal literals, I will do that in a different PR. They're not _exactly_ interdependent 😂

File: ksql-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java (hunk 17 above)
Comment (agavra) [reply ksql-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java]: okay wait nevermind, this is necessary and it does exactly what I thought it does ^

**Comment by blueedgenick** [comment]
Thanks @agavra , looking great! No worries about deferring the implicit naming until later - i'd suggest simply concatenating "levels" of nested-name as the nesting level increases e.g. a generated name for a second-level field might be `FIRST_LEVEL_NAME + COL_0`, which may even turn out to be `KSQL_COL_0_COL_0` if you don't name anything. Bonus points if the fields _inside_ the struct get names with a `FIELD_0` suffix instead of `COL_0`.

And yes, it would be awesome to move map and array creation to work this same way instead of the ...more expedient... way we had done it previously with UDFs 😁 

One thing i realize i forgot to address yesterday is the literal syntax you've proposed for use in the `INSERT...VALUES` statement. It seems like re-stating the name of the field which exists inside the struct should be unnecessary here? e.g. in this example from the original PR description:

> ksql> CREATE STREAM decimals (col0 STRUCT<val DECIMAL(2,1)>) WITH (kafka_topic='decimals', partitions=1, value_format='AVRO');
> ...
> ksql> INSERT INTO decimals (col0) VALUES ({VAL '2.1'});

- here the `VALUES ({VAL '2.1'})` also seems not to fit our desired pattern. Perhaps you already covered this in your changes but I wanted to explicitly call it out to avoid unnecessary revisiting. This could simply be `VALUES (STRUCT( '2.1'))`, right ? The fields of the struct should be in fixed positions so the naming is not needed. Or am I missing something ?

Thanks for getting this one done!

**Comment by agavra** [comment]
> It seems like re-stating the name of the field which exists inside the struct should be unnecessary here

Somewhat true - I think while the schema has ordered fields, the struct does not require ordering. You could decide to map `STRUCT<foo VARCHAR, bar VARCHAR>` to `STRUCT('val1' as BAR, 'val2' as FOO)` (rearranging the fields). So technically a struct value should be paired with a name. You could also entirely omit fields (thus setting them to null) by constructing `STRUCT('val1' as BAR)`.

More importantly, we could add this feature later but it's a non-trivial amount of work. The way `INSERT VALUES` works is to use the same code path as the rest of the expression evaluators; which don't necessarily have schema information.

**Comment by blueedgenick** [comment]
I acknowledge that the current code path for `INSERT...VALUES` is not without it's challenges, perhaps we can at least converge on what we would _like_ to have happen so that anything 'dangling' from this PR can be collected/pasted into a follow-up issue ?
I disagree with using the `STRUCT('val1' as BAR, 'val2' as FOO)` pattern on the grounds that you can't "re-order" the other columns to which you are inserting values in this way, so why special-case just the struct internals ? Also, intuitively, the `x AS y` pattern is for assigning new names to the result of something, not for assigning specific values to specific extant parameters ? Further if you want to set a sub-component of a struct to null while inserting, you should say so explicitly or, at the very least, insert enough commas to skip the sub-fields in question (i strongly prefer explicit nulls though!).

**Comment by agavra** [comment]
> I disagree with using the `STRUCT('val1' as BAR, 'val2' as FOO)` pattern on the grounds that you can't "re-order" the other columns to which you are inserting values in this way, so why special-case just the struct internals ? 

I'm not sure I understand. You can reorder the columns you are inserting values for (`INSERT INTO foo (col1, col2) ...` vs `INSERT INTO foo (col2, col1)`). Also, logically a struct is an unordered mapping of key-value pairs - the fact that the schema has an ordering is an implementation detail and not exposed anywhere else (you cant, for example, do `my_struct[0]`).

> Also, intuitively, the `x AS y` pattern is for assigning new names to the result of something, not for assigning specific values to specific extant parameters ? 

This makes sense, maybe that isn't the right pattern for creating structs then 😉 

> Further if you want to set a sub-component of a struct to null while inserting, you should say so explicitly or, at the very least, insert enough commas to skip the sub-fields in question (i strongly prefer explicit nulls though!).

I disagree with this as well. Explicit nulls are nice, but that's not how `INSERT VALUES` works today with columns either. If I have `CREATE foo (col1 VARCHAR, col2 VARCHAR)` and I `INSERT INTO foo (col1) VALUES( 'foo')` then `col2` will be `null`.

**Comment by blueedgenick** [comment]
>> I disagree with using the STRUCT('val1' as BAR, 'val2' as FOO) pattern on the grounds that you can't "re-order" the other columns to which you are inserting values in this way, so why special-case just the struct internals ?

> I'm not sure I understand. You can reorder the columns you are inserting values for (INSERT INTO foo (col1, col2) ... vs INSERT INTO foo (col2, col1)). Also, logically a struct is an unordered mapping of key-value pairs - the fact that the schema has an ordering is an implementation detail and not exposed anywhere else (you cant, for example, do my_struct[0]).

- not well explained on my part, apologies. I meant to say you don't write something like `insert into foo(col2, col1, col3) values ('x' as col1, 'y' as col3, 1 as col2)` - see how unnatural that looks to use the '..as..' inside of the values section ?

 - re explicit nulls: i know this is how it works today (unfortunately) but i suggest that's only because we don't yet have explicit nullability controls on the schemas i.e. which columns can/cannot be null. I suppose it's something of a gray area how to handle the "sub-fields" of a struct in this case because there's not much precedence for having them in column-based schemas at all. If, for example, you _could_ declare in your schema that struct1 was non-null - which i suppose you are considering to be the case for insert...values anyway in the event that you list struct1 as one of the columns you are inserting to - then what implication does that/should that have for teh sub-components of struct1 ? Do we imagine that in the future they will be individually nullable ? If so, i could be convinced that you may want to explicitly name the ones to which you are assigining values and omit the others in an insert statement, otherwise they have to be declared as null. Either way, i'd argue that the ability to omit them and have them default to null is an optimization of sorts (you're thinking of it like an overloaded object constructor, i suspect ;) ), which we could layer in later as sugar. WDYT ?

**Comment by agavra** [comment]
I don't think your first bullet is a fair comparison 😂If we wanted to take your example, it would be that instead of `INSERT INTO foo (col1, col2) VALUES ('foo1', 'foo2');` you could write `INSERT INTO foo VALUES ('foo1' AS col1, 'foo2' AS col2);`. What makes the example you gave unnatural is not (in my opinion) the `.. as ..` syntax but rather needing to specify the columns twice. The meta-point here is that when you're creating a row in a table, you do need to specify the column names unless you want to specify _all_ columns in order (which I'm proposing is an extension of this work).

With regards to nulls, I think the example I'm thinking of is not really an overloaded constructor but rather the `new Struct(schema).put(fieldX, valueX).put(fieldY, valueY)...` pattern. The schema defines which fields exist, but whether or not you decide to put a value in a struct is orthogonal to the schema.

As a side note, I think we've come to an agreement about non insert-values syntax. I also think it's nice to have the same syntax for insert-values and CSAS statements. That way you can think of an `INSERT VALUES` statement as creating a struct independent of the column schema, and then fitting it into the column.

**Comment by big-andy-coates** [comment]
Hummm...

The concern tha `... AS ... ` should be for aliasing, not passing parameters is valid.  I think there is some precedent in other SQL systems to use syntax similar to `STRUCT(foo => 'x', bar => col0)` or `STRUCT(foo:= 'x', bar:= col0)`.  For example, see postgres 'Using Named Notation' here: https://www.postgresql.org/docs/9.1/sql-syntax-calling-funcs.html.

`INSERT VALUES` matches what most, if not all, SQL systems I've used in the past in the way it doesn't require values to be provided for any nullable or defaulted column.  IMHO, I think it's fine / preferable to have `STRUCT()` follow the same pattern.  I'm also with @agavra in that the order should not be important when named notation is used. However, we could allow non-named notation as well, and in which case the field order can be the same as the STRUCT's definition, (which has order in its field names). Here we can either require explicit `null`s or use the double comma pattern @blueedgenick mentioned.  `INSERT VALUES` supported something similar to named and non-named notation, depending on if the set of columns is provided or not.  If I had to choose I'd go with named first.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java (hunk 5 above)
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java]: I'll name it `getStructSchemaName` and make those corresponding changes

File: ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 18]
```diff
@@ -320,6 +323,19 @@ public Void visitSubscriptExpression(
...
+      final Builder builder = SqlStruct.builder();
+
+      for (Entry<String, Expression> field : exp.getStruct().entrySet()) {
+        process(field.getValue(), context);
+        builder.field(field.getKey(), context.getSqlType());
+      }
+
+      context.setSqlType(builder.build());
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java]: For the first point I think it's okay as you suggested. The coercion happens in the `INSERT INTO`. For the second point, I agree it is better to maintain ordering even if it isn't exposed in KSQL - I have changed all map usages to `List<Field>`

File: ksql-execution/src/test/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitorTest.java [hunk 19]
```diff
@@ -93,11 +95,13 @@
   @Before
   public void init() {
     AtomicInteger funCounter = new AtomicInteger();
+    AtomicInteger structCounter = new AtomicInteger();
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-execution/src/test/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitorTest.java]: this is correctly named! this is actually tracking the number of structs created, and generating the _schema_ name based on that. I've clarified the other places to be `structSchema`, but I stand by my original naming.

File: ksql-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json [hunk 20]
```diff
@@ -272,6 +272,30 @@
...
+      "name": "should handle empty struct expressions",
+      "statements": [
+        "CREATE STREAM TEST (val STRUCT<foo DECIMAL(2, 1), `bar` ARRAY<VARCHAR>>) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "INSERT INTO TEST (val) VALUES (STRUCT());"
+      ],
+      "inputs": [
+      ],
+      "outputs": [
+        {"topic": "test_topic", "key": null, "value": {"VAL": {"FOO": null, "bar": null}}}
+      ]
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json]: We do have explicit tests for this in the `DefaultSqlValueCoercerTest` and tests for usage. I don't think it adds value to have them here in the RQTT test (as you mentioned, I added the String->Decimal coercion to roughly test that this is happening). I can change that to BIGINT/INTEGER if you feel strongly.

[the remark this answers is not in the document]
Comment (agavra) [reply ksql-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json]: Actually, I'll just go ahead and do this

**Comment by big-andy-coates** [comment]
I think there's still the outstanding issue of syntax. As per https://github.com/confluentinc/ksql/pull/4120#issuecomment-566030675, I think its worth considering not using `AS`, which is for aliasing. Rather than `STRUCT('x' AS foo, col0 AS bar)` I propose: `STRUCT(foo:= 'x', bar:= col0)`

**Comment by derekjn** [comment]
> Rather than STRUCT('x' AS foo, col0 AS bar) I propose: STRUCT(foo:= 'x', bar:= col0)

Big +1 from me on this. As @big-andy-coates alluded to, with `STRUCT('x' AS foo, col0 AS bar)` we'd be repurposing the meaning of `AS`, and it's not immediately clear which side of the `AS` is what (`Struct` key or value). PostgreSQL uses `:=` for named parameters, and I believe Oracle uses `=>`. The `:=` operator is more intuitive to me personally, as it represents assignment in pretty much all contexts whereas `=>` can mean different things.

**Comment by agavra** [comment]
@derekjn @big-andy-coates I've made the change to use `:=`!

File: ksql-functional-tests/src/test/resources/query-validation-tests/create-struct.json [hunk 21]
```diff
@@ -0,0 +1,61 @@
+{
+  "comments": [
+    "Tests covering the use of the array returning UDFs."
+  ],
+  "tests": [
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/create-struct.json]: Can we add some negative test cases? e.g. what happens if there are two fields with the same name, or some name but different case? etc. etc.

Also, a test for case-sensitive field names?

Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/create-struct.json]: I guess the tests are awkwardly split between files - i'll add more here
